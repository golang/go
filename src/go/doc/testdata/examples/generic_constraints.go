// Copyright 2021 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package p_test

import (
	"fmt"
	"time"
)

type C1 interface {
	string | int
}

type C2 interface {
	M(time.Time)
}

type G[T C1] int

func g[T C2](x T) {}

type Tm int

func (Tm) M(time.Time) {}

type Foo int

func Example() {
	fmt.Println("hello")
}

func ExampleGeneric() {
	var x G[string]
	g(Tm(3))
	fmt.Println(x)
}
