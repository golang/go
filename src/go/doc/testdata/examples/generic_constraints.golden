-- .Play --
package main

import (
	"fmt"
)

func main() {
	fmt.Println("hello")
}
-- Generic.Play --
package main

import (
	"fmt"
	"time"
)

type C1 interface {
	string | int
}

type C2 interface {
	M(time.Time)
}

type G[T C1] int

func g[T C2](x T) {}

type Tm int

func (Tm) M(time.Time) {}

func main() {
	var x G[string]
	g(Tm(3))
	fmt.Println(x)
}
